enhancements/testing/openshift-tests-extension.md

```diff
@@ -0,0 +1,1202 @@
+is non-trivial to implement, so to simplify the creation of extensions, a Go module &
+repository will be maintained by TRT: https://github.com/openshift-eng/openshift-tests-extension .
+
+Vendoring this module and following its integration guide will provide component test authors
+the majority of the logic necessary to fulfill the test extension interface. It will
+also provide a means of centrally defining new testing requirements. As component owners re-vendor
+the module, they would be required to implement those new testing requirements
+for the module to compile and run a test extension binary successfully.
+
+During a run of the origin test framework (`openshift-tests`), the framework will:
+
+1. Discover available extension binaries.
+2. Discover metadata and component configuration options for those extensions (running `info` verb and parsing standard
+   out).
+3. Discover tests available from those extensions (running `list` verb and parsing standard out).
+4. Plan an appropriate and efficient testing strategy based on the framework invocation.
+5. Execute the tests using subsequent invocations of the discovered binaries (`run-test` invocations).
+6. Collect the result and output of the test invocations and integrate it into overall test suite results.
+
+#### Binary Discovery
+
+The existence of test extension binaries can be registered one of two ways by test authors.
+
+##### OpenShift Payload Extension Binaries
+
+For OpenShift payload components contributors can advertise the existence of an extension binary
+by adding information (the imagestream tag for the OCP payload component and the path to the binary
+within their image) to a simple registry datastructure in github.com/openshift/origin.
+
+##### Non-Payload Extension Binaries
+
+For non-payload components, contributors must advertise their extension binary for `openshift-tests`
+to discover. This is accomplished by creating an ImageStream / ImageStreamTag with a special label:
+
+`testextension.redhat.io/component=<product-name>-<component-name>`
+
+An annotation on the ImageStreamTag will then identify the binary to extract and (optional)
+arguments to pass ahead of the extension interface verbs.
+`testextension.redhat.io/binary=<binary-path>.gz [--argument=value]`
+
+Well-behaved operators should gate the creation of this imagestream/label/annotation on the existence of the
+`TestExtensionAdmission` custom resource definition (the code should not attempt to
+inspect `TestExtensionAdmissions` instances -- just check for the CRD itself so that operators
+don't liter clusters with ImageStreams only applicable for testing in a production environment).
+
+Optional Operator authors must ensure that the image carrying the extension binary
+is identified in their ClusterServiceVersion (CSV) so that tools like `oc-mirror`
+will copy image(s) bearing extension binaries to disconnected clusters.

```

Comment:
We also must ensure that all images used by the extension e2e tests are mirrored by `./openshift-tests images --upstream --to-repository=xyz`.

I could be wrong here, I think this list of images seems to be currently hardwired in openshift-tests binary during build. There is already some wording about adding new images to the list [here](https://github.com/openshift/origin/tree/master/test/extended/util/image), it will need to be either much stricter or we would need to get the list of images from extension binaries too.

Reply:
Justin and I discussed this at one point, but I guess the outcome didn't end up in the enhancement. I think the outcome of that was new images would just need a separate PR to origin to add it.  It's low frequency enough that it shouldn't be too disruptive.

We need to solve the helper problem, though.